### ISSUE TITLE
AWSMobileAnalyticsBufferedReader crash

### ISSUE THREAD

**Issue by stepanhruda** [aws/aws-sdk-ios#151]
Hi, we are getting crashes in production from the SDK and they don't touch any of our code.

The only feature we're using is AWS Mobile Analytics, I assume this crash happens when the SDK is submitting new events up to the server. I'm pretty positive the crashes are a race condition and cannot be reproduced deterministically.

```
Application Specific Information:
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[__NSCFString appendString:]: nil argument'

Last Exception Backtrace:
0   CoreFoundation                       0x000000018487e59c __exceptionPreprocess + 132
1   libobjc.A.dylib                      0x0000000194fc80e4 objc_exception_throw + 56
2   CoreFoundation                       0x000000018487e4dc +[NSException raise:format:] + 124
3   CoreFoundation                       0x0000000184828b94 mutateError + 184
4   OurFramework                         0x000000010028e8dc -[AWSMobileAnalyticsBufferedReader readLine:withError:] (AWSMobileAnalyticsBufferedReader.m:95)
5   OurFramework                         0x00000001002a7478 -[AWSFileEventIterator hasNext] (AWSMobileAnalyticsFileEventStore.m:324)
6   OurFramework                         0x00000001002926bc __72-[AWSMobileAnalyticsDefaultDeliveryClient attemptDeliveryUsingPolicies:]_block_invoke (AWSMobileAnalyticsDefaultDeliveryClient.m:211)
7   Foundation                           0x00000001857374a8 __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ + 12
8   Foundation                           0x0000000185688c34 -[NSBlockOperation main] + 92
9   Foundation                           0x00000001856785bc -[__NSOperationInternal _start:] + 632
10  Foundation                           0x000000018573a20c __NSOQSchedule_f + 224
11  libdispatch.dylib                    0x000000019560d36c _dispatch_client_callout + 12
12  libdispatch.dylib                    0x00000001956174c0 _dispatch_queue_drain + 1212
13  libdispatch.dylib                    0x0000000195610474 _dispatch_queue_invoke + 128
14  libdispatch.dylib                    0x0000000195619224 _dispatch_root_queue_drain + 660
15  libdispatch.dylib                    0x000000019561a75c _dispatch_worker_thread3 + 104
16  libsystem_pthread.dylib              0x00000001957e92e4 _pthread_wqthread + 812
17  libsystem_pthread.dylib              0x00000001957e8fa8 start_wqthread + 0
```

To give you more context, another thread's trace at that time might be relevant as well (this thread did not crash, it was the above one that did):

```
0   libsystem_kernel.dylib               0x000000019574f078 __psynch_cvwait + 8
1   Foundation                           0x0000000185735c04 -[__NSOperationInternal _waitUntilFinished:] + 136
2   Foundation                           0x00000001856db008 -[NSOperationQueue waitUntilAllOperationsAreFinished] + 284
3   OurFramework                         0x0000000100291aec -[AWSMobileAnalyticsDefaultDeliveryClient waitForDeliveryOperations] (AWSMobileAnalyticsDefaultDeliveryClient.m:122)
4   OurFramework                         0x000000010029f5a8 __63-[AWSMobileAnalyticsDefaultSessionClient waitForSessionTimeout]_block_invoke (AWSMobileAnalyticsDefaultSessionClient.m:297)
5   OurFramework                         0x000000010029fde4 -[AWSMobileAnalyticsDelayedBlock execute] (AWSMobileAnalyticsDelayedBlock.m:79)
6   OurFramework                         0x000000010029f650 __63-[AWSMobileAnalyticsDefaultSessionClient waitForSessionTimeout]_block_invoke186 (AWSMobileAnalyticsDefaultSessionClient.m:304)
7   OurFramework                         0x00000001002aaf8c __64-[AWSMobileAnalyticsIOSLifeCycleManager executeBackgroundTasks:]_block_invoke70 (AWSMobileAnalyticsIOSLifeCycleManager.m:148)
8   libdispatch.dylib                    0x000000019560d3ac _dispatch_call_block_and_release + 20
9   libdispatch.dylib                    0x000000019560d36c _dispatch_client_callout + 12
10  libdispatch.dylib                    0x000000019561940c _dispatch_root_queue_drain + 1148
11  libdispatch.dylib                    0x000000019561a75c _dispatch_worker_thread3 + 104
12  libsystem_pthread.dylib              0x00000001957e92e4 _pthread_wqthread + 812
13  libsystem_pthread.dylib              0x00000001957e8fa8 start_wqthread + 0
```

They are both non-main threads.

**Comment by yosuke-matsuda** [comment]
Thank you for the report. We will take a look. Could you share some more details on your environment including:

* How often you encounter the issue
* Version of the AWS SDK
* Version of Xcode
* Version of iOS
* Device type

**Comment by stepanhruda** [comment]
> How often you encounter the issue

It just started happening in a newly released version. I discovered that it currently only happens to a single user, but it happens to them regularly (couple of times a day).

> Version of the AWS SDK

2.0.12

> Version of Xcode

6.1.1

> Version of iOS

iOS 8.1.3

> Device type

iPhone 6

**Comment by yosuke-matsuda** [comment]
Thank you. We will look into it.

**Comment by yyzhu** [comment]
We've identified the bug and the fix is on the way. Thank you.

**Comment by stepanhruda** [comment]
Thanks, I appreciate it!

**Comment by yosuke-matsuda** [comment]
We've released the AWS Mobile SDK for iOS 2.0.17, which addresses the issue. Thanks.
